ops/charm.py

```diff
@@ -128,6 +128,10 @@ class ActionEvent(EventBase):
     params: Dict[str, Any]
     """The parameters passed to the action."""
 
+    def __init__(self, handle: 'Handle', id: str):
```

Comment:
This could be `Optional[str]` and default to `None`, which would provide backwards compatibility. It seems unlikely that much code would be creating an `ActionEvent`, but it is possible (I'll check scenario, for example).

It would be nice if it didn't have to be, so that we don't end up in another situation where charms need to have a bunch of `assert event.id is not None` statements when in production `.id` is actually never None.

Reply:
Indeed, scenario does fail. I can hack a pass with:

```diff
diff --git a/scenario/runtime.py b/scenario/runtime.py
index 489169b..1760200 100644
--- a/scenario/runtime.py
+++ b/scenario/runtime.py
@@ -194,6 +194,7 @@ class Runtime:
             "JUJU_DISPATCH_PATH": f"hooks/{event.name}",
             "JUJU_MODEL_NAME": state.model.name,
             "JUJU_ACTION_NAME": action_name,
+            "JUJU_ACTION_UUID": "1",
             "JUJU_MODEL_UUID": state.model.uuid,
             "JUJU_CHARM_DIR": str(charm_root.absolute()),
             # todo consider setting pwd, (python)path
```

Although I'd have to figure out what the proper fix would be (presumably not always a fixed ID).

There could be other cases where there's an `ActionEvent` created/emitted manually. Once I get my "test lots of charms" script fully working I could test with that, but I'm not sure if it would check enough. It could default to `""` or some other string, so that the type is at least correct? Perhaps with a deprecation warning? I don't love that, though.

Reply:
Per discussion, for backwards-compatibility, yeah, let's make the `__init__` signature take `id: Optional[str] = None` as a keyword arg, but then have the attribute be type `str` (non-optional) with default value `''`.